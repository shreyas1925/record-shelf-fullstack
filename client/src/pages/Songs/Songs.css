@import url('https://fonts.googleapis.com/css2?family=Poppins:wght@700&display=swap');
.song--container{
    /* height:100vh; */
    display: flex;
    flex-direction: column;
    align-items: center;
    
    /* justify-content: center; */
}

.grey--container{
    background-color: #F5F5F5;
    width: 900px;
    /* height: 100vh; */
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    padding-left: 100px;
    padding-right: 100px;
    padding-top: 30px;
}


.grey--container .container{
    width: 100%;
    /* margin: auto;s */
    /* height: 100vh; */
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    /* margin-bottom: 20px; */
    /* margin-top: 20px; */
}

.grey--container .mini--header{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 660px;
    font-weight: 900px;
    font-family: 'Poppins', sans-serif;
}

.grey--container .mini--header img{
    width: 30px;
    height: 30px;
    margin-top: 20px;
    margin-right:20px;
}

.grey--container .mini--header h1{
    margin-top: 20px;
    margin-left: 20px;
}

@media screen and (max-width: 868px) {
*{
    overflow-x: none;
}
    .grey--container .mini--header{
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        width: 100%;
        font-weight: 900px;
        font-family: 'Poppins', sans-serif;
    }
    
    .grey--container .mini--header img{
        width: 30px;
        height: 30px;
        margin-top: 30px;
        margin-bottom: 10px;
        margin-right: 5px;
    }
    
    .grey--container .mini--header h1{
        margin-top: 20px;
        margin-right: 90px;
        margin-left: -25px;
    }
    .grey--container{
        width: 100%;
    }

}