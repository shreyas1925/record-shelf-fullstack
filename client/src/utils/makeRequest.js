// import axios from "axios";


// const makeRequest = async (
//   apiEndPoint,
//   dynamicConfig = {},
// ) => {
//   try{
//       const requestDetails = {
//         baseURL: BACKEND_URL,
//         url: apiEndPoint.url,
//         method: apiEndPoint.method,
//         ...dynamicConfig,
//     };
//     const { data } = await axios(requestDetails);
//     return data;
//   }catch(e)
//   {
//     console.log(e);
//   }
// };

// export default makeRequest;


const BACKEND_URL = "http://localhost:8080/";
const makeRequest = async (path, options) => {
    try{
        options.headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'Authorization': 'Bearer QWlzaHdhcnlhIE4='
        };
        const result = await fetch(`${BACKEND_URL}${path}`,options);
        const {data} = await result.json();
        return data
    } catch(err){
        return err;
    }
};

export default makeRequest;