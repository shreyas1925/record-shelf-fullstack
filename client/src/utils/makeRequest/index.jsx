
// const makeRequest = async (path, options) => {
//     try{
//         options.headers = {
//             'Content-Type': 'application/json',
//             'Accept': 'application/json',
//             'Authorization': 'Bearer QWlzaHdhcnlhIE4='
//         };
//         const result = await fetch(`${BACKEND_URL}${path}`,options);
//         const {data} = await result.json();
//         return data
//     } catch(err){
//         return err;
//     }
// };

// export default makeRequest;

import axios from 'axios';
const BACKEND_URL = 'http://localhost:8080/';
const makeRequest = async (apiEndPoint, dynamicConfig, navigate) => {
    try {
        const requestDetails = {
            baseURL: BACKEND_URL,
            url: apiEndPoint.url,
            method: apiEndPoint.method,
            headers: {
                authorization: 'Bearer QWlzaHdhcnlhIE4=',
            },
            ...dynamicConfig,
        };
        const { data } = await axios(requestDetails);
        return data;
    } catch (error) {
        const errorCode = error.response?.status;
        if (errorCode) {
            navigate(`/error/${errorCode}`);
        } else {
            navigate('/error');
        }
    }
};
  
export default makeRequest;