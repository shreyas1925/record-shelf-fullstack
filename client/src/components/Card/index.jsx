import React, {useEffect, useState } from "react";
import "./Card.css";
import redHeart from "../../assets/heart-red.svg";
import blackHeart from "../../assets/heart-gray.svg";
import makeRequest from "../../utils/makeRequest";
import { GET_SONG_LIKE_BY_ID, UPDATE_SONG_LIKE_BY_ID } from "../../constants/apiEndPoints";

const Card = ({ song }) => {
  
  const [countInc, setCountInc] = useState(0);
  const [likeInc, setLikeInc] = useState(false);

  useEffect(()=>{
    makeRequest(GET_SONG_LIKE_BY_ID(song.id), {})
        .then((response) => {
          setCountInc(response.data.count)
          setLikeInc(response.data.like)
        })
        .catch((e) => {
          // setError(e.message);
      });
  })
  
  const likedImage = likeInc ? redHeart : blackHeart;

  const handleLikes = async () => {
    try{
      await makeRequest(UPDATE_SONG_LIKE_BY_ID(song.id),{ data: { like:!likeInc } });
      setCountInc(likeInc ? countInc - 1 : countInc + 1);
      setLikeInc(!likeInc);
    } catch(e){
      // 
    }
  }
  return (
    <div className="container--card" data-testid='card'>
      <img alt="card" src={song.imageUrl} />
      <div className="card--footer">
        <div className="card--details">
          <h3>{song.name}</h3>
          <p>{song.artist.name}</p>
        </div>
        <div className="red--heart">
          <img src={likedImage} alt="heart" onClick={handleLikes}/>
          <p>{countInc}</p>
        </div>
      </div>
    </div>
  );
};

export default Card;
