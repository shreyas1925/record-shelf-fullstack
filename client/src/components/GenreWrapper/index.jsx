import React from "react";
import Card from "../Card";

const GenreWrapper = ({images,types,text,likes,counts}) => {
  return (
    <div data-testid="genre-wrapper">
      <div className="genre--type">
        <img src={images} alt="genre" />
        <button>{text}</button>
      </div>

      {types.map((song, index) => {
        return <Card song={song} likes={likes[index]} counts={counts[index]} />;
      })}
    </div>
  );
};

export default GenreWrapper;
