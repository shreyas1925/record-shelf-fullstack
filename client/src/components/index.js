import Card from "./Card";
import Navbar from "./Navbar";
import Wrapper from "./Wrapper";

export { Card, Navbar , Wrapper};