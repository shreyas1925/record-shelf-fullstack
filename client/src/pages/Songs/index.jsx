import React, { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import "./Songs.css";
import { Card } from "../../components";
import iconGenre from "../../assets/icon-genre.svg";
import makeRequest from "../../utils/makeRequest";
import { GET_SONGS } from "../../constants/apiEndPoints";


const Songs = () => {

  const [songs, setAllSongs] = useState([]);
  const navigate = useNavigate()
  useEffect(() => {
    makeRequest(GET_SONGS, {},navigate)
      .then((response) => {
        setAllSongs(response.data);
      })
  }, []);

  return (
    <>
    <div className='song--container' data-testid='songs'>
        <div className='grey--container'>
          <div className="mini--header">
            <h1>all songs</h1>
            <Link to="/genre" data-testid='genre'>
              <img src={iconGenre} alt="heart" />
            </Link>
          </div>
          <div className="container" >
           {songs.map((song)=><Card data-testid='card' song={song} />)}
          </div>
      </div>
    </div>
    </>
  );
};

export default Songs;
