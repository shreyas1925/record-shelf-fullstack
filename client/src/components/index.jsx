import Card from "./Card";
import Navbar from "./Navbar";
import Wrapper from "./Wrapper";
import GenreWrapper from "./GenreWrapper";

export { Card, Navbar , Wrapper,GenreWrapper};