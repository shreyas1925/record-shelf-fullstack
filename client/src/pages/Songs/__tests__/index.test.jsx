import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Songs from "..";
import makeRequest from "../../../utils/makeRequest";
import { Card } from "../../../components";

jest.mock("../../../utils/makeRequest");

const mockData = {
  data: [
    {
      id: "45e1d753-2986-43cb-9b9d-30c370056319",
      name: "This Love",
      genre: {
        id: "128aa7f8-c943-48ce-b352-7edd26fa4c6e",
        name: "Pop",
      },
      artist: {
        id: "496b0a85-2bfa-45bc-8d0f-57fe0ce55708",
        name: "Maroon 5",
      },
      imageUrl:
        "https://i.scdn.co/image/ab67616d0000b27317b3850d758fff5a2301e537",
      publishedAt: "2002-06-25T00:00:00",
    },
  ],
};
const mockSong = {
  id: "45e1d753-2986-43cb-9b9d-30c370056319",
  name: "This Love",
  genre: {
    id: "128aa7f8-c943-48ce-b352-7edd26fa4c6e",
    name: "Pop",
  },
  artist: {
    id: "496b0a85-2bfa-45bc-8d0f-57fe0ce55708",
    name: "Maroon 5",
  },
  imageUrl: "https://i.scdn.co/image/ab67616d0000b27317b3850d758fff5a2301e537",
  publishedAt: "2002-06-25T00:00:00",
};

describe("Song Snapshot", () => {
  it("should render forming a snapshot", () => {
    makeRequest.mockResolvedValue(mockData);
    const { asFragment } = render(
      <BrowserRouter>
        <Songs />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Cards length", () => {
  it("should display number of cards", async () => {
    makeRequest.mockResolvedValue(mockData);
    render(
      <BrowserRouter>
        <Songs />
      </BrowserRouter>
    );
    await waitFor(() => {
      expect(screen.getAllByTestId("card").length).toEqual(1);
    });
  });
});

describe("Songs", () => {
  it("should render without crashing", async() => {
    makeRequest.mockResolvedValue(mockData);
    render(
      <BrowserRouter>
        <Songs />
      </BrowserRouter>
    );
    await waitFor(()=>{
      const SongsElement = screen.getByTestId("songs");
      expect(SongsElement).toBeInTheDocument();
    })
    
  });

  it("should display data after rendering", async () => {
    makeRequest.mockResolvedValue(mockData);
    render(
      <BrowserRouter>
        <Songs />
      </BrowserRouter>
    );

    await waitFor(() => {
    const SongsElement = screen.getByText("all songs", { exact: false });
    expect(SongsElement).toBeInTheDocument();
    })
  });

  it("should navigate to songs page after clicking 'sync'", async() => {
    makeRequest.mockResolvedValue(mockData);
    render(
      <BrowserRouter>
        <Songs />
      </BrowserRouter>
    );
    await waitFor(() => {
    const linkElement = screen.getByTestId("genre");
    expect(linkElement).toHaveAttribute("href", "/genre");
    })
  });
});
