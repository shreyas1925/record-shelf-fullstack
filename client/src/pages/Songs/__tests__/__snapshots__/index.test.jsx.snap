// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song Snapshot should render forming a snapshot 1`] = `
<DocumentFragment>
  <div
    class="song--container"
    data-testid="songs"
  >
    <div
      class="grey--container"
    >
      <div
        class="mini--header"
      >
        <h1>
          all songs
        </h1>
        <a
          data-testid="genre"
          href="/genre"
        >
          <img
            alt="heart"
            src="icon-genre.svg"
          />
        </a>
      </div>
      <div
        class="container"
      />
    </div>
  </div>
</DocumentFragment>
`;
